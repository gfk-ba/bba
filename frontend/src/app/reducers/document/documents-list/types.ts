import { KindOption } from '../models/document.class';

export type DocumentsListFilter = Readonly<{
  project: string;
  kind: KindOption | 'all';
  component: string;
}>;
