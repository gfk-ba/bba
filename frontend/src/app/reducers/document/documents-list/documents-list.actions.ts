import { ApiGetRequest } from '@dcs/ngx-tools';
import { Action } from '@ngrx/store';
import { documentsListManager } from './documents-list.manager';
import { DocumentsListFilter } from './types';

export enum DocumentsListActionTypes {
  StartPolling = '[Documents List] Start Polling',
  StopPolling = '[Documents List] Stop Polling',
  SetFilter = '[Documents List] Set Filter',
  ResetFilter = '[Documents List] Reset Filter',
}

export class FetchDocumentsList extends ApiGetRequest {
  constructor() {
    super('documents', documentsListManager.actions.fetch.base, documentsListManager.schema);
  }
}

export class PollDocumentsListStart implements Action {
  public readonly type = DocumentsListActionTypes.StartPolling;
}

export class PollDocumentsListStop implements Action {
  public readonly type = DocumentsListActionTypes.StopPolling;
}

export class SetDocumentsListFilter implements Action {
  public readonly type = DocumentsListActionTypes.SetFilter;
  constructor(public payload: DocumentsListFilter) {}
}

export class ResetDocumentsListFilter implements Action {
  public readonly type = DocumentsListActionTypes.ResetFilter;
}

export type DocumentsListActions =
  | FetchDocumentsList
  | PollDocumentsListStart
  | PollDocumentsListStop
  | SetDocumentsListFilter
  | ResetDocumentsListFilter;
