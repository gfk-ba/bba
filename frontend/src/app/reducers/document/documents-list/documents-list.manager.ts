import { createSelector } from '@ngrx/store';
import { schema } from 'normalizr';
import { DocumentsListFilter } from './types';
import { State } from '../../index';
import { IProject } from '../../project/models/project.class';
import { projectsListStateSelector } from '../../project/projects-list/projects-list.manager';
import { currentDocumentManager } from '../current-document/current-document.manager';
import { Document, IDocument } from '../models/document.class';
import {
  crudCollectionManagerFactory,
  INormalizedCollectionState,
  generateNormalizedCollectionState,
} from '@dcs/ngx-tools';

export interface IDocumentsListState extends INormalizedCollectionState {
  entities: { documents: { [key: string]: IDocument }; projects: { [key: string]: IProject } };
  filter: DocumentsListFilter;
}

const initialState: IDocumentsListState = {
  ...generateNormalizedCollectionState<IDocumentsListState>({
    documents: {},
    projects: {},
  }),
  filter: { component: '', kind: 'all', project: '' },
};

const documentsSchema = new schema.Array(currentDocumentManager.schema);

const pureDocumentsListStateSelector = (state: State): IDocumentsListState => {
  return state.documents.documentsList;
};

const documentsListStateSelector = createSelector(
  [pureDocumentsListStateSelector, projectsListStateSelector],
  (documentsState, projectsState) => {
    return {
      ...documentsState,
      entities: {
        documents: documentsState.entities.documents,
        projects: projectsState.entities.projects,
      },
    };
  }
);

export const documentsListManager = crudCollectionManagerFactory(
  'Documents List',
  'documents',
  initialState,
  currentDocumentManager.actions,
  documentsListStateSelector,
  documentsSchema,
  Document
);
