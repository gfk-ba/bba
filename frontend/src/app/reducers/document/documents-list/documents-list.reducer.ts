import { DocumentsListActions, DocumentsListActionTypes } from './documents-list.actions';
import { documentsListManager, IDocumentsListState } from './documents-list.manager';

export const documentsList = (
  state: IDocumentsListState = documentsListManager.initialState,
  action: DocumentsListActions
): IDocumentsListState => {
  switch (action.type) {
    // overwrite or add any actions here, just default redux
    case documentsListManager.actions.fetch.start:
      return { ...state, loading: true };

    case documentsListManager.actions.fetch.success:
      return { ...documentsListManager.reducer(state, action), filter: state.filter };

    case DocumentsListActionTypes.SetFilter:
      return { ...state, filter: action.payload };

    case DocumentsListActionTypes.ResetFilter:
      return { ...state, filter: documentsListManager.initialState.filter };
  }

  return documentsListManager.reducer(state, action);
};
