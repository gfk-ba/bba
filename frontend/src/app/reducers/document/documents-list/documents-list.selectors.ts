import { createSelector } from '@ngrx/store';
import { documentsListManager, IDocumentsListState } from './documents-list.manager';
import { State } from '../..';

export const filterSelector = (state: State) =>
  (documentsListManager.selectors.subState(state) as IDocumentsListState).filter;

export const filteredDocumentsListSelector = createSelector(
  [documentsListManager.selectors.collection, filterSelector],
  (documents, filter) => {
    return documents
      .filter(
        doc =>
          filter.kind === 'all' || doc.kind.toLocaleLowerCase() === filter.kind.toLocaleLowerCase()
      )
      .filter(doc => {
        if (!filter.project) {
          return true;
        }

        return doc.project.name.toLocaleLowerCase().includes(filter.project.toLocaleLowerCase());
      })
      .filter(doc => {
        if (!filter.component) {
          return true;
        }

        return (
          doc.description.toLocaleLowerCase().includes(filter.component.toLocaleLowerCase()) ||
          doc.hash.toLocaleLowerCase().includes(filter.component.toLocaleLowerCase())
        );
      });
  }
);
