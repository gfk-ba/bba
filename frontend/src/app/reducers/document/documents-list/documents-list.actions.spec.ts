import { FetchDocumentsList } from './documents-list.actions';

describe('document list actions', () => {
  describe('FetchDocumentsList', () => {
    it('builds the correct API request action', () => {
      expect(new FetchDocumentsList()).toMatchSnapshot();
    });
  });
});
