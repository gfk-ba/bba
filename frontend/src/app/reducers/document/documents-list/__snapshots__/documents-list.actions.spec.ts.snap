// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document list actions FetchDocumentsList builds the correct API request action 1`] = `
FetchDocumentsList {
  "payload": Object {
    "cancel": undefined,
    "handlers": "[Documents List] Fetch",
    "normalizrSchema": ArraySchema {
      "schema": EntitySchema {
        "_getId": [Function],
        "_idAttribute": [Function],
        "_key": "documents",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "schema": Object {
          "links": Array [
            EntitySchema {
              "_getId": [Function],
              "_idAttribute": "id",
              "_key": "documents",
              "_mergeStrategy": [Function],
              "_processStrategy": [Function],
              "schema": Object {},
            },
          ],
          "project": EntitySchema {
            "_getId": [Function],
            "_idAttribute": [Function],
            "_key": "projects",
            "_mergeStrategy": [Function],
            "_processStrategy": [Function],
            "schema": Object {},
          },
        },
      },
    },
    "rawDataProcessor": undefined,
    "request": Object {
      "method": "GET",
      "url": "documents",
    },
  },
  "type": "[ApiRequest] Execute API Request",
}
`;
